### PR TITLE
Fix for issue 34185 ovs fail-mode

### LINKED ISSUE #34185: openvswitch_bridge requires the fail_mode parameter
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
openvswitch_bridge

##### ANSIBLE VERSION
```
ansible 2.4.2.0
  config file = /Users/arata/.ansible.cfg
  configured module search path = [u'/Users/arata/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/arata/.pyenv/versions/2.7.11/lib/python2.7/site-packages/ansible
  executable location = /Users/arata/.pyenv/versions/2.7.11/bin/ansible
  python version = 2.7.11 (default, Feb 24 2016, 16:09:27) [GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)]
```

##### CONFIGURATION
nothing with "ansible-config dump --only-changed"

##### OS / ENVIRONMENT
N/A

##### SUMMARY

ansible-doc openvswitch_bridge says that:
```
- fail_mode
        Set bridge fail-mode. The default value (None) is a No-op.
        (Choices: secure, standalone)[Default: None]
        version_added: 2.0
```
However, if the target bridge is set fail mode (either standalone or secure), openvswitch_bridge failed with complaining the fail_mode parameter is not passed.
As the doc says omitting fail_mode means no-op, the module should exit successfully (without changing the fail mode).

##### STEPS TO REPRODUCE
First, create a ovs bridge and set some fail mode (not necessary to use ansible).
```
$ ansible -i hosts host1 --become -m openvswitch_bridge -a 'bridge=ovs1'
```

Then, run openvswitch_bridge to the bridge without fail_mode.
```
ansible -i hosts host1 --become -m openvswitch_bridge -a 'bridge=ovs1'
```

##### EXPECTED RESULTS
Since fail_mode is not passed, the module doesn't care fail mode, just confirms the bridge exists and exits successfully.

##### ACTUAL RESULTS
failed with an error message saying fail_mode must be passed, as follows:
```
$ ansible -i hosts host1 --become -m openvswitch_bridge -a 'bridge=ovs1'
host1 | FAILED! => {
    "changed": false, 
    "cmd": "/usr/bin/ovs-vsctl -t 5 set-fail-mode ovs1 None", 
    "msg": "ovs-vsctl: fail-mode must be \"standalone\" or \"secure\"", 
    "rc": 1, 
    "stderr": "ovs-vsctl: fail-mode must be \"standalone\" or \"secure\"\n", 
    "stderr_lines": [
        "ovs-vsctl: fail-mode must be \"standalone\" or \"secure\""
    ], 
    "stdout": "", 
    "stdout_lines": []
}
```

### PR DESCRIPTION
**Pull request by ashish-billore**
This patch fixes issue #34185
This is related to ovs fail-mode when there is already a fail-mode on the ovs and appears in 'have' and unspecified or none in the request ('want'), as per the default in ovs.

Signed-off-by: Ashish Billore <ashish.billore@gmail.com>

##### SUMMARY

This is related to ovs fail-mode when there is already a fail-mode on the ovs and appears in 'have' and unspecified or none in the request ('want'), as per the default in ovs.

https://github.com/ansible/ansible/issues/34185

Fixes #34185
##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

openvswitch_bridge_module
https://docs.ansible.com/ansible/2.6/modules/openvswitch_bridge_module.html

##### ANSIBLE VERSION

```paste below
ansible 2.4.0.0 and above
```

##### ADDITIONAL INFORMATION

This patch adds a check to see if there is already fail-mode set on the ovs and the request specified has the fail-mode or not. If the fail-mode is specified in the request it works as earlier, just apply the fail-mode in 'want'. If there is no fail-mode specified in the 'want' or the request and if there is already a fail-mode set on the ovs, it resets the fail-mode on the ovs bridge, which is the default in ovs.

Please refer: #34185

```paste below

Before the patch is applied:

[root@osc]# ansible master -i inventory.yaml -m openvswitch_bridge -a "bridge=br-int state=present"
128.90.0.54 | FAILED! => {
    "changed": false, 
    "cmd": "/usr/bin/ovs-vsctl -t 5 set-fail-mode br-int None", 
    "failed": true, 
    "msg": "ovs-vsctl: fail-mode must be \"standalone\" or \"secure\"", 
    "rc": 1, 
    "stderr": "ovs-vsctl: fail-mode must be \"standalone\" or \"secure\"\n", 
    "stderr_lines": [
        "ovs-vsctl: fail-mode must be \"standalone\" or \"secure\""
    ], 
    "stdout": "", 
    "stdout_lines": []
}

After applying the path:

[root@osc]# ansible master -i inventory.yaml -m openvswitch_bridge -a "bridge=br-int state=present"
128.90.0.54 | SUCCESS => {
    "changed": true, 
    "commands": [
        "/usr/bin/ovs-vsctl -t 2 del-fail-mode br-int"
    ], 
    "failed": false
}
[root@osc]#
```

### CODE REVIEW

**Comment by ashish-billore** [comment]
Does this fix look reasonable to be merged or there are any suggestions?

**Comment by no2a** [comment]
@ashish-billore: "No-op" in the documentation string of fail_mode means "leave it untouched", IMO. So when fail_mode is not specified, neither set-fail-mode nor del-fail-mode should be performed.

**Comment by ashish-billore** [comment]
@no2a: Thanks for the suggestion.

I can think of following cases that need to be handled in existing logic:

Already fail_mode is set ('have') on the bridge and the request ('want') has:
1. Unspecified (None): In this case we can go with "No-op" with just retaining the existing fail_mode.
2. User wants to clear the existing 'fail_mode' using 'del-fail-mode'. 
Thanks!

**Comment by ashish-billore** [comment]
@no2a: In above patch, both the cases are now handled.

Please have a look and share your thoughts.

Thanks!

File: lib/ansible/modules/network/ovs/openvswitch_bridge.py
Comment (no2a) [inline lib/ansible/modules/network/ovs/openvswitch_bridge.py:138]: Clearing fail-mode is a new feature. Could you update description of the fail_mode in DOCUMENTATION?

**Review by no2a — CHANGES_REQUESTED** [review]
The code is LGTM. DOCUMENTATION also should be updated though.

**Comment by ashish-billore** [comment]
@no2a Thanks for checking this patch.
I'll perhaps separate the fix for this issue and the new feature of clearing fail-mode in 2 patches and submit it with doc update as well. This will keep the patch consistent with the issue headline.

I'll submit these 2 patches shortly, thanks!

**Comment by ashish-billore** [comment]
@no2a I have reverted the new feature of clearing off the fail_mode in this new patch. Please see if this is fine for fixing this issue.
I'll submit another separate patch for Clearing off the fail_mode feature with updated documentation.
Thanks!

**Comment by ashish-billore** [comment]
@no2a New Feature for clearing off the fail_mode with doc update is here:
https://github.com/ansible/ansible/pull/45469/commits/4bfd24b698068729af9ad4a8baafebf70a44094f

Please have a look, thanks!

**Review by gdpak — CHANGES_REQUESTED** [review]
fail_mode=None should be no_op for backward compatability

**Comment by ashish-billore** [comment]
"fail_mode=None" is as mentioned in the documentation no-op already. As can be seen by this check in the patch:

Line 129: "if wanted_fail_mode is not None:"

https://github.com/ansible/ansible/pull/45469/commits/125893ba03257f96c0f295d9938043aa4ffad334

Also, please note:
There are 2 patches here:
1. https://github.com/ansible/ansible/pull/45469/commits/125893ba03257f96c0f295d9938043aa4ffad334 --> this fixes the issue #34185 

2. https://github.com/ansible/ansible/commit/4bfd24b698068729af9ad4a8baafebf70a44094f --> This adds the new feature for clearing off the already set fail_mode.

Patch 1 is parent of patch 2.
So, while reviewing these patches, please maintain the parent for 2nd patch.

Thanks!

**Comment by ashish-billore** [comment]
These patches now handle case where fail_mode is passed as 'None'. Please review. Thanks for the help.

[root@osc]# ansible master -i inventory.yaml -m openvswitch_bridge -a "bridge=br-int state=present fail_mode=None"
128.90.0.54 | SUCCESS => {
    "changed": false, 
    "commands": [], 
    "failed": false
}
[root@osc]# ansible master -i inventory.yaml -m openvswitch_bridge -a "bridge=br-int state=present fail_mode=None"
128.90.0.54 | SUCCESS => {
    "changed": false, 
    "commands": [], 
    "failed": false
}
[root@osc]# ansible master -i inventory.yaml -m openvswitch_bridge -a "bridge=br-int state=present fail_mode="
128.90.0.54 | SUCCESS => {
    "changed": true, 
    "commands": [
        "/usr/bin/ovs-vsctl -t 2 del-fail-mode br-int"
    ], 
    "failed": false
}
[root@osc]# ansible master -i inventory.yaml -m openvswitch_bridge -a "bridge=br-int state=present fail_mode=None"
128.90.0.54 | SUCCESS => {
    "changed": false, 
    "commands": [], 
    "failed": false
}
[root@osc]# ansible master -i inventory.yaml -m openvswitch_bridge -a "bridge=br-int state=present fail_mode=secure"
128.90.0.54 | SUCCESS => {
    "changed": true, 
    "commands": [
        "/usr/bin/ovs-vsctl -t 2 set-fail-mode br-int secure"
    ], 
    "failed": false
}
[root@osc]# ansible master -i inventory.yaml -m openvswitch_bridge -a "bridge=br-int state=present"
128.90.0.54 | SUCCESS => {
    "changed": false, 
    "commands": [], 
    "failed": false
}

**Comment by ashish-billore** [comment]
@no2a @gdpak @webknjaz Can you please check if commit 56c9cf8d70e833f53f93f9983ba53ed9a10bbc82 looks fine now?
Test results are as above: https://github.com/ansible/ansible/pull/45469#issuecomment-425765820

File: lib/ansible/modules/network/ovs/openvswitch_bridge.py
Comment (gdpak) [inline lib/ansible/modules/network/ovs/openvswitch_bridge.py:128]: @ashish-billore  It looks much better now. Thanks for making changes. One comment here is passing empty value to ansible module arg does not look good. Probably we can pass something as 'clear' to convey user intention to clear previously set fail_mode. Also fail_mode is a choice parameter in doc string so it has to be a choice parameter in code as well. This also need fixing. I could test following code to be working as expected. please take a look
```diff
--- a/lib/ansible/modules/network/ovs/openvswitch_bridge.py
+++ b/lib/ansible/modules/network/ovs/openvswitch_bridge.py
@@ -53,9 +53,11 @@ options:
               To  clear all external-ids pass an empty value.
     fail_mode:
         version_added: 2.0
-        choices : [secure, standalone]
+        choices : [secure, standalone, clear]
         description:
             - Set bridge fail-mode. The default value (None) is a No-op.
+              Omitting this parameter is a No-op.
+              To clear previously set fail_mode, pass it as 'clear'.
     set:
         version_added: 2.3
         description:
@@ -123,12 +125,23 @@ def map_obj_to_commands(want, have, module):
             commands.append(command)
     else:
         if have:
-            if want['fail_mode'] != have['fail_mode']:
-                templatized_command = ("%(ovs-vsctl)s -t %(timeout)s"
-                                       " set-fail-mode %(bridge)s"
-                                       " %(fail_mode)s")
-                command = templatized_command % module.params
-                commands.append(command)
+            wanted_fail_mode = want['fail_mode']
+            have_fail_mode = have['fail_mode']
+            if wanted_fail_mode != have_fail_mode:
+                if wanted_fail_mode is not None:
+                    if wanted_fail_mode != 'clear':
+                        # only set the fail-mode if specified and not empty
+                        templatized_command = ("%(ovs-vsctl)s -t %(timeout)s"
+                                               " set-fail-mode %(bridge)s"
+                                               " %(fail_mode)s")
+                    else:
+                        # if fail-mode is specified, but empty it means user
+                        # wants to reset the fail-mode, so clear it off
+                        templatized_command = ("%(ovs-vsctl)s -t %(timeout)s"
+                                               " del-fail-mode %(bridge)s")
+
+                    command = templatized_command % module.params
+                    commands.append(command)
 
             if want['external_ids'] != have['external_ids']:
                 templatized_command = ("%(ovs-vsctl)s -t %(timeout)s"
@@ -233,7 +246,8 @@ def main():
         'state': {'default': 'present', 'choices': ['present', 'absent']},
         'timeout': {'default': 5, 'type': 'int'},
         'external_ids': {'default': None, 'type': 'dict'},
-        'fail_mode': {'default': None},
+        'fail_mode': {'default': None, 'choices': ['secure', 'standalone',
+            'clear']},
         'set': {'required': False, 'default': None}
     }
 
```

**Review by gdpak — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/ovs/openvswitch_bridge.py
Comment (ashish-billore) [reply lib/ansible/modules/network/ovs/openvswitch_bridge.py:128]: thanks @gdpak looks good to me, minor updates to the comment in the above else check for clear:
```suggestion
                    else:
                       # if fail-mode is specified as 'clear' means user
                       # wants to reset the fail-mode, so clear it off
```

**Review by ashish-billore — COMMENTED** [review]
Looks fine to me, minor updates in comments for the 'clear' off else check

**Comment by ashish-billore** [comment]
@no2a @gdpak @gundalow thanks for your interest in this long open PR and providing valuable comments. If it looks fine to cover needed cases, please advise to take it forward.
Thanks!

File: lib/ansible/modules/network/ovs/openvswitch_bridge.py
Comment (webknjaz) [inline lib/ansible/modules/network/ovs/openvswitch_bridge.py:131]: Please merge this clause with previous if-clause using `and`, this will reduce the whole block indentation. Less nesting is better.

**Comment by gdpak** [comment]
@ashish-billore  Can you please commit changes, which we agreed to this PR . Let's get it in devel.

**Comment by ashish-billore** [comment]
@webknjaz @gdpak I will update the patch and submit again.

**Comment by abenokraitis** [comment]
@ashish-billore Any updates? Thanks! needs_info

**Comment by ashish-billore** [comment]
@abenokraitis Sorry for the delay on this, will submit updated patch shortly.
Thanks!

**Comment by Qalthos** [comment]
Hey, we can't proceed on this without the changes. If you could update the PR, we can move forward on this. Thanks.

needs_info

**Comment by pabelanger** [comment]
recheck

**Comment by pabelanger** [comment]
recheck

**Review by pabelanger — CHANGES_REQUESTED** [review]
I'd like to see us update tests/integration/targets/openvswitch_bridge and add a new test to validate this code.

We now have pre-merge tests running from zuul.ansible.com, so help ensure it is properly working before we merge this.

**Comment by ashish-billore** [comment]
recheck

**Comment by ashish-billore** [comment]
recheck

**Comment by ashish-billore** [comment]
recheck

**Comment by danielmellado** [comment]
Thank you very much for your interest in Ansible. This plugin is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/openvswitch.openvswitch
Please re-submit this PR in the above repository.
If you have further questions please stop by IRC or the mailing list:

IRC: #ansible-network on irc.freenode.net
mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Pull request closed without merging**
